teamscale-maven-plugin/src/main/java/com/teamscale/maven/TeamscaleMojoBase.java

```diff
@@ -0,0 +1,131 @@
+ */
+public abstract class TeamscaleMojoBase extends AbstractMojo {
+
+	/**
+	 * The URL of the Teamscale instance to which the recorded coverage will be uploaded.
+	 */
+	@Parameter()
+	public String teamscaleUrl;
+
+	/**
+	 * The Teamscale project to which the recorded coverage will be uploaded
+	 */
+	@Parameter()
+	public String projectId;
+
+	/**
+	 * The username to use to perform the upload. Must have the "Upload external data" permission for the {@link
+	 * #projectId}. Can also be specified via the Maven property {@code teamscale.username}.
+	 */
+	@Parameter(property = "teamscale.username")
+	public String username;
+
+	/**
+	 * Teamscale access token of the {@link #username}. Can also be specified via the Maven property {@code
+	 * teamscale.accessToken}.
+	 */
+	@Parameter(property = "teamscale.accessToken")
+	public String accessToken;
+
+	/**
+	 * You can optionally use this property to override the code commit to which the coverage will be uploaded. Format:
+	 * {@code BRANCH:UNIX_EPOCH_TIMESTAMP_IN_MILLISECONDS}
+	 * <p>
+	 * If no end commit is manually specified, the plugin will try to determine the currently checked out Git commit.
+	 */
+	@Parameter
+	public String endCommit;
+
+	/**
+	 * You can optionally use this property to override the revision to which the coverage will be uploaded.
+	 * If no revision is manually specified, the plugin will try to determine the current git revision.
+	 */
+	@Parameter
+	public String revision;
+
+	/**
+	 * Whether to skip the execution of this Mojo.
+	 */
+	@Parameter(defaultValue = "false")
+	public boolean skip;
+
+	/**
+	 * The running Maven session. Provided automatically by Maven.
+	 */
+	@Parameter(defaultValue = "${session}")
+	public MavenSession session;
+
+	/**
+	 * The resolved commit, either provided by the user or determined via the GitCommit class
+	 */
+	protected String resolvedCommit;
+
+	/**
+	 * The resolved revision, either provided by the user or determined via the GitCommit class
+	 */
+	protected String resolvedRevision;
+
+	/**
+	 * Sets the <code>resolvedEndCommit</code> and <code>resolvedRevision</code>, if not provided, via the GitCommit class
+	 * @see GitCommit
+	 */
+	protected void resolveEndCommit() throws MojoFailureException {
+		if (StringUtils.isNotBlank(endCommit)) {
+			resolvedCommit = endCommit;
+			if (StringUtils.isNotBlank(revision)) {
+				resolvedRevision = revision;
+				return;
+			}
+		}
+		Path basedir = session.getCurrentProject().getBasedir().toPath();
+		try {
+			GitCommit commit = GitCommit.getGitHeadCommitDescriptor(basedir);
+			if (StringUtils.isBlank(resolvedCommit)) {
+				resolvedCommit = commit.branch + ":" + commit.timestamp;
+			}
+			resolvedRevision = commit.sha1;
+		} catch (IOException e) {
+			throw new MojoFailureException("You did not configure an <endCommit> in the pom.xml" +
+					" and I could also not determine the checked out commit in " + basedir + " from Git", e);
```

Comment:
I think it's odd that this is written in the first person here: 
```suggestion
			throw new MojoFailureException("There is no <endCommit> configured in the pom.xml" +
					" and it was not possible to determine the checked out commit in " + basedir + " from Git", e);
```